File: algorithms/W/Walking Robot Simulation II/Walking Robot Simulation II.cpp
class Robot {
private:
    int width, height, perimeter;
    int dir = 0;
    int x = 0, y = 0;
    map<int, string> mp = {{0, "East"}, {1, "North"}, {2, "West"}, {3, "South"}};

public:
    Robot(int width, int height) {
        this->width = width;
        this->height = height;
        this->perimeter = 2 * (height + width) - 4;
    }
    
    void step(int num) {
		if(!num) return;
        num = num % perimeter;
        if(x == 0 && y == 0 && num == 0) dir = 3;
        while(num--){
            if(x == 0 && y == 0) dir = 0;
            else if(x == width-1 && y == 0) dir = 1;
            else if(x == width-1 && y == height-1) dir = 2;
            else if(x == 0 && y == height-1) dir = 3;
            
            if(dir == 0) x += 1; 
            if(dir == 1) y += 1;
            if(dir == 2) x -= 1;
            if(dir == 3) y -= 1;
        }
        
    }
    
    vector<int> getPos() {
        return {x, y};
    }
    
    string getDir() {
        return mp[dir];
    }
};

File: scripts/algorithms/W/Walking Robot Simulation II/Walking Robot Simulation II.py
class Robot:

    def __init__(self, width: int, height: int):
        self.perimeter = 2*width + 2*(height - 2)
        self.pos = 0
        self.atStart = True

        self.bottomRight = width - 1
        self.topRight = self.bottomRight + (height - 1)
        self.topLeft = self.topRight + (width - 1)

    def step(self, num: int) -> None:
        self.atStart = False
        self.pos = (self.pos + num) % self.perimeter

    def getPos(self) -> List[int]:
        if 0 <= self.pos <= self.bottomRight:
            return [self.pos, 0]

        if self.bottomRight < self.pos <= self.topRight:
            return [self.bottomRight, self.pos - self.bottomRight]

        if self.topRight < self.pos <= self.topLeft:
            return [self.bottomRight - (self.pos - self.topRight), self.topRight - self.bottomRight]
        
        return [0, self.topRight - self.bottomRight - (self.pos - self.topLeft)]

    def getDir(self) -> str:
        if self.atStart or 0 < self.pos <= self.bottomRight:
            return 'East'

        if self.bottomRight < self.pos <= self.topRight:
            return 'North'

        if self.topRight < self.pos <= self.topLeft:
            return 'West'
        
        return 'Sout